File: Data-structure/Tree/tree_order.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

root = Node("F",
            Node("B",
                 Node("A"),
                 Node("D",
                      Node("C"), Node("E"))
                 ),
            Node("G",
                 None,
                 Node("I", Node("H"))
                 )
            )

_pre = []
_in = []
_post = []

# 전위순회
def preorder(node:Node):
    if node is None:
        return
    _pre.append(node.val)
    preorder(node.left)
    preorder(node.right)

# 중위순회
def inorder(node:Node):
    if node is None:
        return
    inorder(node.left)
    _in.append(node.val)
    inorder(node.right)

# 후위순회
def postorder(node:Node):
    if node is None:
        return
    postorder(node.left)
    postorder(node.right)
    _post.append(node.val)

preorder(root)
inorder(root)
postorder(root)

print(*_pre)  # F B A D C E G I H
print(*_in)   # A B C D E F G H I
print(*_post) # A C E D B H I G F

File: Data-structure/Tree/treeOrder.js
class Node {
  constructor(value) {
    this.value = value;
    this.left = this.right = null;
  }
}

class Tree {
  constructor(node) {
    this.root = node;
  }

  levelOrder() {
    const result = [];
    const q = [];
    q.push(this.root);
    while (q.length > 0) {
      const currentNode = q.shift();
      result.push(currentNode.value);
      if (currentNode.left) q.push(currentNode.left);
      if (currentNode.right) q.push(currentNode.right);
    }
    return result;
  }

  preOrder(node = this.root) {
    const preOrderResult = [];

    function _preOrder(node) {
      if (!node) return;
      preOrderResult.push(node.value);
      _preOrder(node.left);
      _preOrder(node.right);
    }
    _preOrder(node);
    return preOrderResult;
  }

  inOrder(node = this.root) {
    const inOrderResult = [];

    function _inOrder(node) {
      if (!node) return;
      _inOrder(node.left);
      inOrderResult.push(node.value);
      _inOrder(node.right);
    }
    _inOrder(node);
    return inOrderResult;
  }

  postOrder(node = this.root) {
    const postOrderResult = [];

    function _postOrder(node) {
      if (!node) return;
      _postOrder(node.left);
      _postOrder(node.right);
      postOrderResult.push(node.value);
    }
    _postOrder(node);
    return postOrderResult;
  }
}

const tree = new Tree(new Node(9));
tree.root.left = new Node(3);
tree.root.right = new Node(8);
tree.root.left.left = new Node(2);
tree.root.left.right = new Node(5);
tree.root.right.right = new Node(7);
tree.root.left.right.right = new Node(4);

console.log(tree.levelOrder()); // [ 9, 3, 8, 2, 5, 7, 4 ]

console.log(tree.preOrder()); // [ 9, 3, 2, 5, 4, 8, 7 ]

console.log(tree.inOrder()); // [ 2, 3, 5, 4, 9, 8, 7 ]

console.log(tree.postOrder()); // [ 2, 4, 5, 3, 7, 8, 9 ]
